source/compilation/Compilation.cpp

```diff
@@ -448,6 +448,15 @@ static void reportRedefinition(const Scope& scope, const T& newSym, const U& old
     }
 }
 
+template<typename T, typename U>
+static void reportDuplicateDefinition(const Scope& scope, const T& newSym, const U& oldSym) {
```

Comment:
Since this basically just copies the reportRedefinition method, how about reusing that and making it take a DiagCode argument that you pass on to addDiag(). Then each call site can specify whether it's the warning or error version just by changing the DiagCode it passes.